### ISSUE TITLE
API google.pubsub.v1.Publisher exceeded 5000 milliseconds when running on Cloud Run

### ISSUE THREAD

**Issue by buffolander** [googleapis/nodejs-pubsub#1442]
When the image is run locally on Docker the application is able to successfully publish messages to Cloud PubSub.

When the same image is deployed to Cloud Run, it won't publish a single message. All attempts fail with the error:

```
GoogleError: Total timeout of API google.pubsub.v1.Publisher exceeded 5000 milliseconds before any response was received.
    at repeat (/usr/root/node_modules/google-gax/build/src/normalCalls/retries.js:66:31)
    at Timeout._onTimeout (/usr/root/node_modules/google-gax/build/src/normalCalls/retries.js:101:25)
    at listOnTimeout (internal/timers.js:557:17)
    at processTimers (internal/timers.js:500:7)
code 4
```

Sample Code:

```
import { mongoose } from '@shared/connections'
import MongoDbModel from '@shared/models'
import { PubSub } from '@google-cloud/pubsub'

import {
  REMINDER_COPY_SMS = '',
  REMINDER_MAX_DAYS_SINCE_LAST_MSG = 0,
  REMINDER_MAX_DAYS_SINCE_LAST_READING = 0,
  REMINDER_MESSAGE_TEMPLATE = 'reading-reminder',
  TOPICS_MESSAGE_DISPATCH,
  KEYFILE_PATH,
} from '../constants.js'

const pubsubClient = new PubSub({
  projectId: process.env.GOOGLE_CLOUD_PROJECT,
  keyFilename: KEYFILE_PATH,
})

const msMaxSinceLastReading = REMINDER_MAX_DAYS_SINCE_LAST_READING * 24 * 60 * 60 * 1000
const msMaxSinceLastMsg = REMINDER_MAX_DAYS_SINCE_LAST_MSG * 24 * 60 * 60 * 1000

const { model: Patient } = new MongoDbModel(mongoose, 'Patient')

const findInactivePatients = () => Patient.find({
  status: 'ACTIVE',
  $and: [{
    $or: [{
      last_reading_at: { $exists: false },
    }, {
      last_reading_at: { $lt: new Date(Date.now() - msMaxSinceLastReading) },
    }],
  }, {
    $or: [{
      last_reading_reminder_at: { $exists: false },
    }, {
      last_reading_reminder_at: { $lt: new Date(Date.now() - msMaxSinceLastMsg) },
    }],
  }, {
    // Test with known patients only
    $or: [{
      first_name: { $regex: /Bruno/i }, last_name: { $regex: /Soares/i },
    }],
  }],
}, {
  first_name: 1,
  last_name: 1,
  gender: 1,
  last_reading_at: 1,
  phones: 1,
}, {
  lean: true,
})

const updateLastReadingReminderAt = (patient) => Patient.findOneAndUpdate({
  _id: patient._id,
}, {
  last_reading_reminder_at: Date.now(),
})

const createMessagePayload = (patient) => ({
  body: REMINDER_COPY_SMS,
  channels: [{
    name: 'sms',
    contacts: patient.phones.map((phone) => phone.E164),
    specifications: {
      template: REMINDER_MESSAGE_TEMPLATE,
    },
  }],
})

const dispatchMessage = async (patient) => {
  if (!patient.phones || !patient.phones.length) {
    return
  }
  const payload = createMessagePayload(patient)
  console.info('Worker reminder-report-vitals', 'event payload', JSON.stringify(payload))
  try {
    await pubsubClient.topic(TOPICS_MESSAGE_DISPATCH).publishMessage({ json: payload })
    console.info('Worker reminder-report-vitals', 'event published to cloudPubSub')
    pubsubClient.close()
    await updateLastReadingReminderAt(patient)
    console.info('Worker reminder-report-vitals', 'last_reading_reminder_at updated for patient', patient._id)
  } catch (err) {
    console.error(err)
  }
}

const handler = async () => {
  console.info('Worker reminder-report-vitals', 'exectution started')
  try {
    const inactiveList = await findInactivePatients()
    inactiveList.forEach(dispatchMessage)
  } catch (err) {
    console.error(err)
  }
}

export default handler

```

The code hangs after logging this line:
`console.info('Worker reminder-report-vitals', 'event payload', JSON.stringify(payload))`

Sample Event Payload:

```
{"body":"This is {{ORG}}. We haven't been receiving your vitals. Reply \"start\" to report your vitals now. Reply \"stop\" at any time to opt-out from automated reminders.","channels":[{"name":"sms","contacts":["+15555555555"],"specifications":{"template":"reading-reminder"}}]}
```

#### Environment details

  - OS: Linux Alpine
  - Node.js version: 14.16.1
  - `@google-cloud/pubsub` version: 2.18.4

#### Steps to reproduce

  1. Create PubSub topic to receive messages and replace its name in constants.js TOPICS_MESSAGE_DISPATCH.
  2. Create nodejs application image from sample code. For sake of replicating only the issue, database operations may be removed - instead the **sample event payload** may be used.
  3. Run the container locally with Docker.
  4. Push the image to GCP Container Registry
  5. Deploy a Cloud Run service using the image
  6. See timeout erro in Cloud Logs

**Comment by buffolander** [comment]
FYI I experienced the same behavior running on Cloud Functions. Rolling back to version 2.12.0 unblocked me.

**Comment by Sytten** [comment]
Yeah we are experiencing issues on cloud run.

**Comment by maurocolella** [comment]
For what it's worth, I am seeing the same error message when running the emulator locally, which in that case appears to be because the port used by the emulator (random, in the 8000 range) doesn't match the one exported through the `PUBSUB_EMULATOR_HOST` variable.

I'll try a container-based approach, but I suspect a port mismatch.

**Comment by Sytten** [comment]
I don't believe this is the issue in production. Most likely cloud run now kills the network access when the container is inactive (aka when the request has ended) which happens more often on low load. It used to work though, so I believe it might be a change in cloud run infrastructure. I am making sure that all the pubsub calls are awaited properly on my side. You should also set `.topic(topic, { batching: { maxMessages: 1 } })` so that the message is sent right away instead of waiting in a queue.

**Comment by maurocolella** [comment]
For context: I was referring to an issue with ports on the emulator. I have deleted the relevant comment because I also believe it's not what's happening here. Apologies for the confusion.

**Comment by haroldadmin** [comment]
Can confirm, we are facing the same issue on Cloud Functions with `@google-cloud/pubsub` version 2.18.4.

**Comment by gorziza** [comment]
What's new about this case? I have the same problem

**Comment by Sytten** [comment]
We revered to 2.17. I believe the bug was introduced in 2.18.

**Comment by meredithslota** [comment]
@Sytten Just checking — the issue does not reoccur with 2.17, only with 2.18? Which minor version were you using? I do see a change that went out re: publish timeouts, https://github.com/googleapis/nodejs-pubsub/releases/tag/v2.18.0 and then updated in https://github.com/googleapis/nodejs-pubsub/releases/tag/v2.18.3. It looks like @feywind did some investigation in an earlier issue here: https://github.com/googleapis/nodejs-pubsub/issues/1425 and it seems like there was an upstream change we're trying to figure out. Our current believe is that 2.18.3 fixes the issue, but please let me know if this isn't the case.

**Comment by apettiigrew** [comment]
Getting the same issue in google cloud function trigger.

**Comment by gorziza** [comment]
@meredithslota We had the problem with 2.18.3 too

**Comment by causticsudo** [comment]
same here too,  with 2.18.3 and 2.18.3 version

**Comment by apettiigrew** [comment]
> FYI I experienced the same behavior running on Cloud Functions. Rolling back to version 2.12.0 unblocked me.

Tried this solution it did not work for my case, while using a background cloud function. 
```
const { PubSub } = require("@google-cloud/pubsub");
const pubsub = new PubSub();

exports.main = async (message, context) => {
  const data = {
    event: "AccountCreated"
  };

  const pubMessage = Buffer.from(JSON.stringify(data));
  const topic = pubsub.topic(topicName);

  topic.publish(pubMessage);
};
```

```
{
  "name": "sample-pubsub",
  "version": "0.0.1",
  "dependencies": {
    "@google-cloud/pubsub": "^2.12.0"
  }
}
```

**Comment by gorziza** [comment]
@apettiigrew Remove ^... use version only. "@google-cloud/pubsub": "2.17.0"

**Comment by cmagk** [comment]
It occured to me today as well after upgrading from 2.17x, reverted back but why did they introduce 5s timeout?

**Comment by lguendogdu** [comment]
Same here, too. Running 2.18.4 with dockerized node application.

**Comment by Benny739** [comment]
Same problem here with 2.15.1

**Comment by BerangerNt** [comment]
Same here using google cloud functions with @google-cloud/pubsub, reverted to "@google-cloud/pubsub": "2.17.0" and it's working again.

**Comment by BerangerNt** [comment]
I had the same issue using Google Cloud Functions. I was publishing successive messages probably the wrong way :
using "batching" option and publishing them using Promise.all reduced the processing time, it's now under 5000 ms

**Comment by andersondanilo** [comment]
Inside cloud run, i needed to disable the batching (maxMessages = 1), and make sure the promise was awaited before the http response was returned.

**Comment by andersondanilo** [comment]
> @Sytten Just checking — the issue does not reoccur with 2.17, only with 2.18? Which minor version were you using? I do see a change that went out re: publish timeouts, https://github.com/googleapis/nodejs-pubsub/releases/tag/v2.18.0 and then updated in https://github.com/googleapis/nodejs-pubsub/releases/tag/v2.18.3. It looks like @feywind did some investigation in an earlier issue here: #1425 and it seems like there was an upstream change we're trying to figure out. Our current believe is that 2.18.3 fixes the issue, but please let me know if this isn't the case.

Inside cloud run, i needed to disable the batching (maxMessages = 1), and make sure the promise was awaited before the http response was returned.

Note: By default the cloud run indeed dont allocate cpu after the request is finished (https://cloud.google.com/blog/products/serverless/cloud-run-gets-always-on-cpu-allocation)

**Comment by andersondanilo** [comment]
> I don't believe this is the issue in production. Most likely cloud run now kills the network access when the container is inactive (aka when the request has ended) which happens more often on low load. It used to work though, so I believe it might be a change in cloud run infrastructure. I am making sure that all the pubsub calls are awaited properly on my side. You should also set `.topic(topic, { batching: { maxMessages: 1 } })` so that the message is sent right away instead of waiting in a queue.

Inside cloud run, i needed to disable the batching (maxMessages = 1), and make sure the promise was awaited before the http response was returned.

Note: By default the cloud run indeed dont allocate cpu after the request is finished (https://cloud.google.com/blog/products/serverless/cloud-run-gets-always-on-cpu-allocation)

**Comment by minuhariharan** [comment]
We are facing the same issue as well, using pub sub version 2.18.3, is downgrading the only solution?

**Comment by minuhariharan** [comment]
We are seeing this error: Received error while publishing: Total timeout of API google.pubsub.v1.Publisher exceeded 600000 milliseconds before any response was received.

**Comment by cmagk** [comment]
> We are facing the same issue as well, using pub sub version 2.18.3, is downgrading the only solution? 

We only use pubsub in Nodejs for some tests locally. We use the Java sdk on production where it works fine.

I tried deploying it to the gcloud just to see and didn't get the timeout there.I do not know why but I am thinking there is something wrong with the way we have implemented nodejs.

Will do some more tests with some other versions of writing it today or tomorrow.

**Comment by timdauer** [comment]
> We are facing the same issue as well, using pub sub version 2.18.3, is downgrading the only solution?

Same here - we are using the version 2.18.1 though. So I guess downgrading needs to go at least below that - if that helps anyway.

We also opened (yet another) P3 support case at google. They will refer to the product team - lets hope this issue resolves quick.

**Comment by rgmvisser** [comment]
We are seeing the same issue. It doesn't happen on every message, but it seems pretty unreliable. 

Can anyone confirm that downgrading to 2.17 will fix this in the meantime?

**Comment by linde12** [comment]
I think this is kind of related to #847 - there were some PRs (see thread) to solve this issue and i believe that it is now causing this timeout. Previously, pushing e.g. 4k messages of messages resulted in like 30k messages being published instead; now when i try to publish 4k messages around 2k are published then we get this timeout error instead...

**Comment by Fajing** [comment]
I had the same warning in CF log. Rolling back to version 2.12.0 unblocked me.

**Comment by gmcheco** [comment]
Same error. 2.12.0 worked for me.
Thks @buffolander 👍🏼

**Comment by bms13** [comment]
I had to update "@google-cloud/pubsub": "^2.12.0" and  also add "google-gax": "^2.28.1" into package.json to get this to work.

**Comment by feywind** [comment]
@buffolander Just to double check something in the originally posted code:

```
  try {
    await pubsubClient.topic(TOPICS_MESSAGE_DISPATCH).publishMessage({ json: payload })
    console.info('Worker reminder-report-vitals', 'event published to cloudPubSub')
    pubsubClient.close()
    await updateLastReadingReminderAt(patient)
    console.info('Worker reminder-report-vitals', 'last_reading_reminder_at updated for patient', patient._id)
  } catch (err) {
    console.error(err)
  }
```

This probably shouldn't be calling `pubsubClient.close()` unless you don't expect to execute this function again. CF can hold on to container instances it makes for functions, so Node may remain running, and thus `pubsubClient` would still be alive between requests. There _were_ some changes not too long ago with how `close()` handles shutdown, if I recall correctly, so it might be related.

**Comment by feywind** [comment]
I should also mention, you might want to cache the result of `pubsubClient.topic()` since that can be reused for future publishes, and it will make better use of message batching.

**Comment by syJSdev** [comment]
It happens only emulator mode on my side.

Pubsub Receiver Project
firebase-tools: 9.16.0
  @google-cloud/pubsub: 2.16.1

Pubsub Sender Project
@google-cloud/pubsub: 2.18.5

Is this cause of Issue?
@feywind

**Comment by barthje** [comment]
Funny, the change has been reverted in the latest version in this commit: https://github.com/googleapis/nodejs-pubsub/commit/34a4d4aa50c6f07adcec1dcba1ee0eaf941f71c0 a ticket to update the copyright years.

The ticket that it belonged to: https://github.com/googleapis/nodejs-pubsub/pull/1465

**Comment by barthje** [comment]
I can confirm that updating to the latest version solved my problem.

**Comment by cdupuis** [comment]
Everyone affected by this issue should also review https://github.com/googleapis/nodejs-pubsub/issues/1463

**Comment by HappyHappy1996** [comment]
Folks make sure you don't have `PUBSUB_EMULATOR_HOST` environment variable set when your app is running (obviously if you want to use GCP pubsub, not emulator).
In my case, this was the issue, as I had this environment variable equal to `localhost:PORT` which was giving timeout errors all the time when I had expected it to make requests to GCP - after `.env` was excluded from the image we use deploy in GCP the problem is solved.
Cheers

**Comment by raygerami** [comment]
I am getting the same issue of 600000-millisecond timeout when calling pubishMessage locally. I initially had @google-cloud/pubsub: ^2.10.0.

I tried all combinations of upgrading and adding gaxios to my packages.json, nothing helped.
The creation and subscription work fine with the pubsub-client but when pushing the messages it shows the times out error.

I thought it may be an environmental issue, and I changed the code to invoke pure rest-API using JWT client from 'google-auth-library', which uses the same gaxios client internally, and the rest-API works fine.

Not sure, it seems the client is not even trying any HTTP post and gets stuck somewhere.

Thanks

**Comment by maxfrigge** [comment]
The only reliable fix that worked for us was deploying to CloudRun with `--no-cpu-throttling`.

**Comment by maxfrigge** [comment]
@Sytten the endpoint will have quite a bit of traffic later on so it's OK for us.

We have tried using `{ batching: { maxMessages: 1 } })` and it worked most of the time, but occasionally failed as well.

**Comment by feywind** [comment]
Wow, that commit message for https://github.com/googleapis/nodejs-pubsub/commit/34a4d4aa50c6f07adcec1dcba1ee0eaf941f71c0 was really not great. It looks like the protos actually changed to set back the timeout value, and OwlBot just happened to pull them in.

Setting the max message queue to 1 is expensive in other ways, but if you're literally sending one message in a Cloud Function, that might honestly be the best way to handle it. The library is currently designed more with bulk send in mind.

**Comment by growlytics** [comment]
For me this was happening in version 2.17 too.
I needed to create a dummy version in cloud run to restart servers.

**Comment by raashidkhan1** [comment]
I have tried both pubsub 2.17 and gax 2.28.1, but doesn't seem to work for this simple use case-
```
exports.main = async (message, context) => {
  const data = {
    event: "AccountCreated"
  };
  const topicName = 'some valid topic name'
  const pubMessage = Buffer.from(JSON.stringify(data));
  const topic = pubsub.topic(topicName);

  topic.publish(pubMessage);
};
```

Keep getting the same error when the await on publishMessage hangs for a bit and crashes-
```
GoogleError: Total timeout of API google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.
    at repeat (/Users/rocktech/LeidenUniv/sorting-as-a-service/server/node_modules/@google-cloud/pubsub/node_modules/google-gax/build/src/normalCalls/retries.js:66:31)
    at Timeout._onTimeout (/Users/rocktech/LeidenUniv/sorting-as-a-service/server/node_modules/@google-cloud/pubsub/node_modules/google-gax/build/src/normalCalls/retries.js:101:25)
    at listOnTimeout (node:internal/timers:564:17)
    at process.processTimers (node:internal/timers:507:7) {
  code: 4
}
```

**Comment by linde12** [comment]
> I have tried both pubsub 2.17 and gax 2.28.1, but doesn't seem to work for this simple use case-
> 
> ```
> exports.main = async (message, context) => {
>   const data = {
>     event: "AccountCreated"
>   };
>   const topicName = 'some valid topic name'
>   const pubMessage = Buffer.from(JSON.stringify(data));
>   const topic = pubsub.topic(topicName);
> 
>   topic.publish(pubMessage);
> };
> ```
> 
> Keep getting the same error when the await on publishMessage hangs for a bit and crashes-
> 
> ```
> GoogleError: Total timeout of API google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.
>     at repeat (/Users/user/path/app/server/node_modules/@google-cloud/pubsub/node_modules/google-gax/build/src/normalCalls/retries.js:66:31)
>     at Timeout._onTimeout (/Users/user/path/app/server/node_modules/@google-cloud/pubsub/node_modules/google-gax/build/src/normalCalls/retries.js:101:25)
>     at listOnTimeout (node:internal/timers:564:17)
>     at process.processTimers (node:internal/timers:507:7) {
>   code: 4
> }
> ```

Are you calling it concurrently? I had a similar issue to this post when running om GCP and my issue was that i was sending lots of request concurrently without awaiting the promises, causing my service to scale down. Instead you can do multiple publishes in a single request and await the respones (if doing scale-to-zero)

**Comment by raashidkhan1** [comment]
This is actually running on a local machine, not publishing multiple messages concurrently but sequentially within a loop, the error comes on the first iteration itself.
The code looks similar to
```
const topicName = 'some valid topic name'
const topic = pubsub.topic(topicName);
array.forEach(async (value)=>{
  const pubMessage = Buffer.from(JSON.stringify(value));
  const topic = pubsub.topic(topicName);
  await topic.publish(pubMessage);
}
```

**Comment by RuslanUshakov** [comment]
I face with the same issue while using pubsub emulator. But previously it seemed working:
```
GoogleError: Total timeout of API google.pubsub.v1.Subscriber exceeded 60000 milliseconds before any response was received.
           |     at repeat (/usr/src/app/node_modules/@google-cloud/pubsub/node_modules/google-gax/src/normalCalls/retries.ts:84:23)
           |     at Timeout._onTimeout (/usr/src/app/node_modules/@google-cloud/pubsub/node_modules/google-gax/src/normalCalls/retries.ts:125:13)
           |     at listOnTimeout (node:internal/timers:559:17)
           |     at processTimers (node:internal/timers:502:7) {
           |   code: 4

```

**Comment by feywind** [comment]
Linked to the meta-issue about transport problems: b/242894947

**Comment by JamesTheTerry** [comment]
Not to make this even stranger, I've observed this happening when running Cloud Functions (HTTP CF v1) that are triggered by Cloud Tasks HTTP calls. What's odd is I have many Pub/Sub triggered Cloud Functions (Event driven/background functions, also CF v1) that are part of the same flow, and I have not observed the error in them.

I have been running this using `^3.0.1` across all of them.

--
CCing @campbellbrobbel

**Comment by nous-** [comment]
I get `google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.` off cloud run. Our use is so trivial that I have a hard time believing its something we did.

**Comment by blue-hope** [comment]
I fixed like this:

before:

```js
await publisherClient.publish({
  topic: `projects/${projectId}/topics/${topic}`,
  messages: [
    {
      data: JSON.stringify(body),
    },
  ],
});
```

after:

```js
await publisherClient
  .topic(`projects/${projectId}/topics/${topic}`)
  .publishMessage({
     data: Buffer.from(JSON.stringify(body)),
  });
```

**Comment by nous-** [comment]
@blue-hope Your fix seems like a red-herring. Here is what I'm doing in production and started hitting these errors. Our events/sec on this topic is max 100/s so the use of `orderKey` seems like non issue.

```js
// Where eventsTopic is created before hand as eventsTopic = publisherClient.topic('thetopic')
eventsTopic.publishMessage({
    json: {
      op: 'event',
      this: 'that',
      thenext: 1
    },
    orderingKey: specificKey
  });
  ```

**Comment by dominusLabs** [comment]
Just ran into this issue today. Is there a known fix? The issue only seems to happen when we deploy to our staging servers. Locally, everything works as expected.

**Comment by pdfrod** [comment]
Also ran into this yesterday, but on GKE instead of Cloud Run (I think Cloud Run is basically Knative on top of GKE).

Any tips to debug this if it happens again?

**Comment by dominusLabs** [comment]
I managed to fix this issue by caching the topic when the HTTP server started and just reusing it. I'm confused as to why this worked though, because even though the old implementation didn't cache the topic, during our tests we were only sending 1 HTTP request to the web server.

**Comment by aviramha** [comment]
Happening here as well.

**Comment by ItaiLeibowitz** [comment]
I get this issue happening quite a bit using Google cloud functions. 

Super simple message subscriptions (using it to call function 2 from function 1 so there's a single subscription between function 1 --> function 2, ), low load (handful of messages / hr). 

For some function connections (e.g., f1->f2), everything works great, messages seem to always get published correctly. 

For other function connections (eg f3->f4) it's super erratic. ~50% of the time, it will fail with this error message (GoogleError: Total timeout of API google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received). The other 40% of the time, the messages are published just fine. And then maybe 10% of cases, the message is published twice...?

**Comment by nous-** [comment]
I'm suspicious this might be related to [ECONNRESET](https://cloud.google.com/run/docs/troubleshooting#connection-reset-by-peer). I was hitting this error for other kinds of connections, moved the CPU to "CPU is always allocated", and no longer get these messages.

**Comment by okaforcj** [comment]
dont know if this helps anyone but in my case, i wasnt validating the pubsub client properly, as i'm using a non-google service to call the publisher
```js
  const pubsub = new PubSub({
        projectId,
        grpc,
        credentials: {
            client_email: context.values.get("GCP_CLIENT_EMAIL"),
            private_key: context.values.get("GCP_PRIVATE_KEY").replace(/\\n/g, '\n'),
        },
    })
```
    
specifically  i didnt add ` .replace(/\\n/g, '\n')` and for 2 days i've been trying to understand why it was doing the timeout because it doesnt error out with authorisation issues.

**Comment by bullishgopher** [comment]
> I fixed like this in version 3.2.0:
> 
> before:
> 
> ```js
> const publisherClient = new PublisherClient({ projectId })
> await publisherClient.publish({
>   topic: `projects/${projectId}/topics/${topic}`,
>   messages: [
>     {
>       data: JSON.stringify(body),
>     },
>   ],
> });
> ```
> 
> after:
> 
> ```js
> const publisherClient = new PubSub({ projectId })
> await publisherClient
>   .topic(`projects/${projectId}/topics/${topic}`)
>   .publishMessage({
>      data: Buffer.from(JSON.stringify(body)),
>   });
> ```

I am on version 3.3.0 and I am having the same issue

**Comment by JamesTheTerry** [comment]
@bullishgopher are you saying that your fix ("after" code snippet) resolved the issue?

**Comment by bullishgopher** [comment]
Sorry @JamesTheTerry 
I mean I have the same issue 

> error - unhandledRejection: GoogleError: Total timeout of API google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.

**Comment by kamalaboulhosn** [comment]
While the symptom of these issues (publisher timeouts) is similar across the cases, it is highly likely the the underlying causes are different and so this GitHub issue now covers a lot of different cases where more user-specific information is needed. If you are still experiencing issues, please create a support case. Thanks!

**Comment by koushil-mankali** [comment]
I got erro "google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.", By increasing gaxOpts timeout value to 100,000 solved our problem.

    const publishOptions = {
      gaxOpts: {
        timeout: 100000,
      },
    };

    const topicChannel = pubsub.topic(TOPIC_ID, publishOptions);

**Comment by bharathjinka09** [comment]
> I got error "google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.", By increasing gaxOpts timeout value to 100,000 solved our problem.
> 
> ```
> const publishOptions = {
>   gaxOpts: {
>     timeout: 100000,
>   },
> };
> 
> const topicChannel = pubsub.topic(TOPIC_ID, publishOptions);
> ```

Thanks @koushil-mankali for your solution. It's working. I am facing a new issue related to google cloud logging. Can you please let me know how to fix the below error. It will be very helpful. Thanks

**Error: Total timeout of API google.logging.v2.LoggingServiceV2 exceeded 60000 milliseconds before any response was received.**

**Comment by davestimpert** [comment]
Does anyone know why it would take over a minute to publish?

**Comment by bharathjinka** [comment]
> Does anyone know why it would take over a minute to publish?

If the queue has a lot of data to process, it will take more time. You can reproduce the issue by load testing with multiple requests. The issue will be fixed by increasing the timeout value to a max value of 540000 i.e. 540 secs.

**Comment by davestimpert** [comment]
Something to consider - we noticed this issue arising from an HttpFunction where we were sending a response before publishing to PubSub. Having awaits after sending a response is contraindicated by google, and when we switched the order to publish first, the issue went away.

**Comment by lemndev** [comment]
Some point to consider about this issue, in case your application is not receiving consistent traffic/load, this error may affect the initial request(s)/action(s) that boot(s) up your cloud run instance(s). With subsequent requests/actions in the same timeframe, you'll notice that this error may not re-occur. 

I cannot explain why it takes that long to publish to pub/sub but increasing the timeout as in [koushil-mankali](https://github.com/koushil-mankali)'s solution [here](https://github.com/googleapis/nodejs-pubsub/issues/1442#issuecomment-1600768399) will save you some headaches.

**Comment by ts-geek22** [comment]
> > I got the error "google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.", By increasing the gaxOpts timeout value to 100,000, we solved our problem.
> > ```
> > const publishOptions = {
> >   gaxOpts: {
> >     timeout: 100000,
> >   },
> > };
> > 
> > const topicChannel = pubsub.topic(TOPIC_ID, publishOptions);
> > ```
> 
> Thanks, @koushil-mankali, for your solution. It's working. I am facing a new issue related to Google Cloud logging. Can you please let me know how to fix the below error? It will be very helpful. Thanks
> 
> **Error: The total timeout of API google.logging.v2.LoggingServiceV2 exceeded 60000 milliseconds before any response was received.**

@bharathjinka09 Did you find any solutions for your last error? I'm facing the same problem, and I have tried several solutions but haven't succeeded.

**Solutions I have tried:**

1. Increased timeout from 60 seconds to 600 seconds in pub/sub client.
2. Increased CPU and RAM for cloud functions.
3. Used batch messaging for efficiency.
4. Create a single topic instance and use it across all messages.

**Comment by kamalaboulhosn** [comment]
Just to reiterate what I said back in June: While the symptom of these issues (publisher timeouts) is similar across the cases, it is highly likely the the underlying causes are different and so this GitHub issue now covers a lot of different cases where more user-specific information is needed. If you are still experiencing issues, please create a support case. Thanks!

Increasing the gax timeout may be a viable solution in some instances, but it is not a general-purpose answer to how to address timeouts in all cases.

We need specific information for each individual case in order to properly diagnose these issues and so a support case is the proper venue for that exchange of information.

**Comment by nsrivastava645-ghl** [comment]
I solved this issue for me by having a single instance of the topic so you can change your code to something like this:

`
const topics = {};
const dispatchMessage = async (patient) => {
  if (!patient.phones || !patient.phones.length) {
    return
  }
  const payload = createMessagePayload(patient)
  console.info('Worker reminder-report-vitals', 'event payload', JSON.stringify(payload))
  try {
    if(!topics[TOPICS_MESSAGE_DISPATCH]){
      topics[TOPICS_MESSAGE_DISPATCH] = pubsubClient.topic(TOPICS_MESSAGE_DISPATCH, {
        batching: {
          maxMessages: 100, // set it to something like 200-300 depending upon how your subscriber is configured.
          maxMilliseconds: 100, // max wait before messages are published to the server (in ms)
        }
      })
    }
    await topics[TOPICS_MESSAGE_DISPATCH].publishMessage({ json: payload })
    console.info('Worker reminder-report-vitals', 'event published to cloudPubSub')
    pubsubClient.close()
    await updateLastReadingReminderAt(patient)
    console.info('Worker reminder-report-vitals', 'last_reading_reminder_at updated for patient', patient._id)
  } catch (err) {
    console.error(err)
  }
}`

Give it a try.

**Comment by ts-geek22** [comment]
> > I have solved this issue. The problem was with environmental variables. Google Cloud functions parse each variable added from the GUI console, and ads escape sequences. So `\n` in my service account's private will be converted to `\\n` resulting in wrong secrets.
> > It's strange that the Pub/Sub SDK does not return proper error messages; it always returns an API timeout, no matter how much time I increase the timeout.
> > Removing escape sequences from the account's private key solves the problem.
> 
> Hi @ts-geek22, were you able to publish messages through the service at all? Or did the error occur after the application had been running for around 1-2 hours?

It's a string parsing issue, so it occurs from the start, as a wrong string is always the wrong string.

**Comment by ts-geek22** [comment]
A timeout error is a general error in the case of cloud publishing, as it throws the same error for multiple reasons and does not provide any specific contextual information.

This is a link to my [Stack Overflow issue](https://stackoverflow.com/q/78086638/16981097), where I have listed a few different solutions I have tried; feel free to give them a try; you might find one that works for you.

**Comment by jdziek** [comment]
Im experiencing a similar issue and a solution that worked for me was to initiate pubsub inside the function and then close the topic. Reason for this was apparently that we kept batching it, resulting in cloud function closing before it had a chance to send it. 

```
    try {
      // WARNING shifting new Pubsub outside of this function and removing close causes pubub time
      const pubSubClient = new PubSub();
      const topic = pubSubClient.topic(topicName, {
        batching: { maxMessages: 1 },
        gaxOpts: {
          timeout: 100000,
        },
      });

      const messageId = await topic.publishMessage({ json }).catch((err) => {
        logger.error(
          `Received error while publishing to ${topicName}: ${err.message}`
        );
        logger.error(err);
        throw new InternalServerError();
      });
      await pubSubClient.close();
      logger.info(
        `Message ${messageId} was published to ${topicName} with message ${JSON.stringify(
          json
        )}`
      );
    } catch (err) {
      logger.error(err);
      throw new InternalServerError();
    }
 }
```

However, somebody changed it according to documentation a while ago and we started having that issue again just recently, so need to retest that. Just throwing this idea out there, maybe it will help somebody. Will try to give an update after more testing. 

In the mentions above, I noticed people talking about importance of awaiting pubsub calls. In our use case we return a response from a server after we call a function that triggers pubsub that is not awaited to pretty much avoid having pubsub fail before we get a chance to respond. However, we await withing the called function. Example


```
async function example () {
....
      initPubsubMessages(
      message1,
      message2,
      ...
      );
return data;
}

async function initPubsubMessages(message1, message2) {

  await Pusub.publish(topic, message1);
  await Pusub.publish(topic, message2);
}

```
We use firebase functions gen 2 which are pretty much google cloud functions at this point. To my knowledge. In my head everything here is correct and until the processes initiated in initPubsubMessages, which has been awaited, should be processed without a hitch. But I'm also aware the cloudrun has some quirks so I thought ill ask for somebody else's opinion on that. Would firebase functions/cloudrun stop all processes as soon the server responds, is my question i guess? 
I am grasping at straws here but I've experienced weirder things using that platform.

**Comment by ts-geek22** [comment]
@jdziek I don't know much about firebase functions, but Cloud function will unallocate memory and CPU as soon as you return response in HTTP cloud function. It assumes that your function's execution is complete as you have returned a response.

**Comment by pdfrod** [comment]
The same issue about Cloud Run was already mentioned here before: https://github.com/googleapis/nodejs-pubsub/issues/1442#issuecomment-1006978912

**Comment by jdziek** [comment]
Was worried that this might be the case. The thing is that this has been happening in places where we await as well. Will add an await just to be safe in this case, but doubt it will solve the issue fully :/ 
Now I just have to hope that hanging pubsub wont prevent response alltogether. 
Thank you.

**Comment by jdziek** [comment]
Was good for about two weeks. Checked we are awaiting everything. Put in tweaks mentioned above. Just had a pubsub error chain this morning. Really at a loss here. Was really hoping that it will work. Any suggestions at this point welcome. Might try to put pubsub back to how it was setup in documentation, maybe this will work this time with proper awaitig.

**Comment by bkovari** [comment]
@jdziek 

> Was good for about two weeks. Checked we are awaiting everything. Put in tweaks mentioned above. Just had a pubsub error chain this morning. Really at a loss here. Was really hoping that it will work. Any suggestions at this point welcome. Might try to put pubsub back to how it was setup in documentation, maybe this will work this time with proper awaitig.

Are you referring to your configuration mentioned earlier?
```
      const topic = pubSubClient.topic(topicName, {
        batching: { maxMessages: 1 },
        gaxOpts: {
          timeout: 100000,
        },
      });
```

**Comment by jdziek** [comment]
Yeah. Noticed though that its happening only in one service now. So currently reviewing the code there. Reverted pubsub implementation to how its shown in the docs and pushed. So hopefully nothing new will appear in two weeks

**Comment by Kripu77** [comment]
> Yeah. Noticed though that its happening only in one service now. So currently reviewing the code there. Reverted pubsub implementation to how its shown in the docs and pushed. So hopefully nothing new will appear in two weeks

Hey @jdziek are you using any other library to connect to any other Google API's?

**Comment by jdziek** [comment]
yeah, no. Just google-cloud ones for firestore, pubsub, and storage.

**Comment by Kripu77** [comment]
> yeah, no. Just google-cloud ones for firestore, pubsub, and storage.

Update those packages to the latest versions; hopefully, that will fix your issue. We faced a similar problem with a few of our services; in our case, it was the Google Secret Manager package. Once we had bumped it to the latest version, that fixed this problem. I believe it's due to the gRPC library that most of the npm packages from Google use for communication with their APIs

**Comment by jdziek** [comment]
Good point. I only updated Pubsub really. Will review it now. Thanks

**Comment by sylvainar** [comment]
Hey, I'm subscribing as well to this issue.

I have the same error in all my cloud run services, everything was fine until a few weeks ago when we started having this issue. We batched to 1, we opened and closed the client for each message, but we're still having it. I'm trying to bump all google dependencies to see if it changes anything, let's keep each other posted!

**Comment by vinay-panwar-04** [comment]
Hi, folks - I am publishing almost 100k messages to a topic somehow after a while of publishing the topic I started getting the "Error: Total timeout of API google. pub-sub.v1.Publisher exceeded 600000 milliseconds before any response was received" and no matter how many timeout values I increased the error will be still there with waiting for that amount of time specified as the timeout value.

**Comment by kdawgwilk** [comment]
@vinay-panwar-04 Same i have the timeout increased on the function and can confirm the timeout on the function is larger than 60000ms and i can also see the acknowledgment deadline for the pubsub subscription is also greater than 60000ms but yet i still see this error
```
Error: Total timeout of API google.pubsub.v1.Publisher exceeded 60000 milliseconds before any response was received.
    at repeat (/workspace/node_modules/google-gax/build/src/normalCalls/retries.js:66:31)
    at Timeout._onTimeout (/workspace/node_modules/google-gax/build/src/normalCalls/retries.js:102:25)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
```

I don't see a way to set the `gaxTimeout` for firebase functions

**Comment by Kripu77** [comment]
> Good point. I only updated Pubsub really. Will review them now. Thanks

Hey @jdziek, did updating the packages fixed the issue for you?

**Comment by sylvainar** [comment]
Hey all,

We had this timeout issue happening randomly on our Cloud Run instances, that could be mitigated by deploying a new revision. I haven't investigated the root cause, but instinctively it looks like when the Cloud Run instance runs for too long, the connection to pubsub stops working, maybe for something related to auth or long-term gax connections.

We finally managed to fix the issue. We needed to upgrade all google packages, including `@google/<something>` packages, `google-auth-library`, `google-gax`, `googleapis-common` and other sub-dependencies.

If (like us) you're using yarn v1, here's the list of commands that can help you:
- `yarn upgrade-interactive`: select all google packages and bump them
- `yarn outdated`: see all packages that have new major versions
- https://www.npmjs.com/package/yarn-deduplicate to remove duplicates in your lockfile (that's the issue we had, we had olders versions of google packages that were still in our lockfile). 
- `yarn why googleapis-common` to check that you have a single version of a lib and that is the latest one

Hope it helped, good luck!

**Comment by jdziek** [comment]
@Kripu77 So far so good. We havnt had issues. Yet. The problem has been so erratic that i cant say for sure yet, but its seems at least like its improved.

**Comment by bkovari** [comment]
With the new configuration the errors seem to be disappeared completely. There were no occurance in the last 2 weeks:
```
    const publishOptions = {
         gaxOpts: {
             timeout: 540000
         },
     };
    const target = pubsub.topic(rule.topic, publishOptions);
```
Previously, I did increase the timeout of the cloud function itself, which did not help.

**Comment by LincolnRychecky** [comment]
I had this issue as well when publishing messages from a V2 node cloud function and I found a solution which resolved my problem. Increasing the timeout as @bkovari recommended prevented the timeout error from being repeated, but there are a few other things to keep in mind when firing large volumes of messages from cloud functions.

I found that batching messages in large numbers led to the node PubSub client hanging in the Cloud Function's execution and not delivering anything. I resolved this by limiting the batch size to one. This may not be ideal for your use case, but for the vast majority it will be fine:

`  const publishOptions = {
    batching: {
      maxMessages: 1,
       maxMilliseconds: 1,
    },
    gaxOpts: {
      timeout: 540000,
    },
  };`
  
  The next thing I can recommend if you are working with JS or TS is resolving the promise from the message send immediately before moving on to your next message send. Allowing numerous unresolved message send promises to build up seems to result in the PubSub client hanging and quietly failing.
  
Here is the JS code that worked for me:
  
 `import {PubSub} from "@google-cloud/pubsub";

export const pubsubService = {
  publishMessage,
};

// initialize pubsub client with token
const pubSubClient = new PubSub();

interface Attributes {
  [k: string]: string;
}

/**
 * Publishes a pubsub message to a topic.
 *
 * @param {string} topicNameOrId - The name or id of the topic you would like to publish to.
 * @param {string} data - The data to publish to the topic
 * @throws {Error} Will throw an error if the message cannot be published.
 */
async function publishMessage(topicNameOrId: string, data: string) {
  const publishOptions = {
    batching: {
      maxMessages: 1,
       maxMilliseconds: 1,
    },
    gaxOpts: {
      timeout: 540000,
    },
  };

  // Add two custom attributes, origin and username, to the message
  const customAttributes: Attributes = {
    "Content-Type": "application/json",
  };

  const dataBuffer = Buffer.from(data);
  try {
    const messageId = await pubSubClient
      .topic(topicNameOrId, publishOptions)
      .publishMessage({
        data: dataBuffer,
        attributes: customAttributes,
      });
    console.log(`Message ${messageId} published.`);
  } catch (error) {
    console.error(
      `Received error while publishing: ${(error as Error).message}`
    );
    process.exitCode = 1;
  }
}
`

In your cloud function you use the function like so and you should have no more (or at the very least less) problems with message delivery:

`        await pubsubService.publishMessage(pubsubTopic, jsonString);
`
